src/zjs_spi.c

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) 2017, Intel Corporation.
+
+// Zephyr includes
+#ifndef ZJS_LINUX_BUILD
+#include <zephyr.h>
+#include <spi.h>
+#include <misc/util.h>
+#endif
+#include <string.h>
+#include <stdlib.h>
+
+// ZJS includes
+#include "zjs_spi.h"
+#include "zjs_util.h"
+#include "zjs_buffer.h"
+#include "zjs_callbacks.h"
+#include "zjs_error.h"
+
+#define SPI_BUS "SPI_"
+#define MAX_SPI_BUS 127
+#define MAX_READ_BUFF 1000
+#define MAX_DIR_LEN 13
+
+static jerry_value_t zjs_spi_prototype;
+
+static void zjs_spi_callback_free(void *native)
+{
+    // requires: handle is the native pointer we registered with
+    //             jerry_set_object_native_handle
+    //  effects: frees the spi item
+    spi_handle_t *handle = (spi_handle_t *)native;
+    handle->closed = true;
+    zjs_free(handle);
+}
+
+static const jerry_object_native_info_t spi_type_info =
+{
+   .free_cb = zjs_spi_callback_free
+};
+
+static ZJS_DECL_FUNC(zjs_spi_transceive)
+{
+    // requires: Writes and reads from the SPI bus, takes one argument
+    //           arg[0] - Buffer to write to the SPI bus
+    //  effects: Writes the buffer to the SPI bus and returns the buffer
+    //           received on the SPI in reply.
+
+    ZJS_VALIDATE_ARGS(Z_NUMBER, Z_OPTIONAL Z_ARRAY Z_STRING, Z_OPTIONAL Z_STRING);
+
+    // No more than 3 args
+    if (argc > 3)
+        return ZJS_STD_ERROR(RangeError, "Invalid args");
+
+    ZJS_GET_HANDLE(this, spi_handle_t, handle, spi_type_info);
+
+
+    if (handle->closed == true) {
+        ZJS_PRINT("SPI bus is closed\n");
+        return jerry_create_null();
+    }
+
+    size_t len;
+    jerry_value_t buffer;
+    zjs_buffer_t *txBuf;
+    zjs_buffer_t *rxBuf;
+    jerry_value_t rxBuf_obj = jerry_create_null();
+    jerry_value_t txBuf_obj;
+
+    if (spi_slave_select(handle->spi_device, jerry_get_number_value(argv[0]))) {
+        return zjs_error("SPI slave select failed\n");
+    }
+
+    // If only a slave is given, this must be a single read or its invalid
+    if (argc == 1) {
+        if (handle->topology != ZJS_TOPOLOGY_SINGLE_READ) {
+            return ZJS_STD_ERROR(RangeError, "Missing args for transceive");
+        }
```

Comment:
"Missing transmit buffer"? (perhaps the message needs to be aligned with other APIs)